### PR TITLE
perl: disallow cc in perl

### LINKED ISSUE #46077: Perl has huge closure in Nixpkgs
## Issue description

```
$ du -sch $(nix-store -qR $(nix-build https://nixos.org/channels/nixpkgs-unstable/nixexprs.tar.xz -A perl --argstr system x86_64-darwin)) | sort -h
0	/nix/store/df6k4mgdjxciy0f637lryp7c9ln7n1m3-cctools-binutils-darwin
12K	/nix/store/2jpbn191ck79zsfcfrwml3zlcxp70c2d-expand-response-params
40K	/nix/store/j3dl6hfba0rqy042lm5815h8r9r8zf51-libffi-3.2.1
52K	/nix/store/nsi1xkmlk3dd51ln3jyv2dmgzpj9miwx-cctools-binutils-darwin-wrapper
76K	/nix/store/2gdwhdzhy4iwkp7fh8v6gy6nxj1zi9pv-clang-wrapper-5.0.2
100K	/nix/store/r3nzm8qdsfb19fys6ah01mjyzyylxwpp-zlib-1.2.11
276K	/nix/store/jl3672s05cp6ga3shjf8ks78gkl3bfll-gnugrep-3.1
296K	/nix/store/llafcziny4071gkahynh9348yqfwvych-compiler-rt-5.0.2-dev
568K	/nix/store/2db7957cxf1g49aay9aiflf52ihbsy95-pcre-8.42
644K	/nix/store/hxvhqnnizi9752q2wq46ymrj0pa9fj87-gmp-6.1.2
668K	/nix/store/jd065p28z3ar0q6nljz479f9mm35v5qn-libc++abi-5.0.2
1.1M	/nix/store/567l21pfpvjb3lr5c6hgshasn589raal-libiconv-osx-10.11.6
1.1M	/nix/store/cblfnvb5rmhd2z231mqasn0brzh1hhv4-bash-4.4-p23
1.7M	/nix/store/zkfqg7icazgz0kj7rff6zdfbcf3v79sa-coreutils-8.29
2.5M	/nix/store/2cf0hbdgyixp2jl3k6l5mdg15pvkn8ah-CF-osx-10.10.5
4.3M	/nix/store/nnlz1ydq8p2iy87b8hh5vlpa4ng7bs8h-compiler-rt-5.0.2
6.9M	/nix/store/qjmqrzk8nhn5maa093fhl3finczz72lw-libc++-5.0.2
7.0M	/nix/store/1j81scbyzgxsv7dnq6naxvi4lb5q8ln9-ICU-osx-10.10.5
7.4M	/nix/store/1ij45avzzzpy965lyf92nrpfnmaxm102-ncurses-6.1
8.1M	/nix/store/a10rrqfq8yz3sdv5vlzvss0b3z8l9vs9-cctools-port-895
9.3M	/nix/store/7zcsy4dhx1zcibmqjjnghcmd296bp4ii-Libsystem-osx-10.11.6
49M	/nix/store/xhir85mv5mfwdkbisnx3bl9h6agjmix4-perl-5.28.0
51M	/nix/store/6dqj75a9lq7kkfzsrf2fx52d9ajj6hvv-llvm-5.0.2-lib
59M	/nix/store/jxkvfglvsik7l0s7ig9wbw8xgznikywd-clang-5.0.2-lib
130M	/nix/store/34iaq9pl11q2hs1pxd0x8hf530hxrqx7-binutils-2.30
399M	/nix/store/q6wsn9mvlyiwg1risizpx8dwaky70szh-clang-5.0.2
738M	total
```

```
$ nix why-depends $(nix-build https://nixos.org/channels/nixpkgs-unstable/nixexprs.tar.xz -A perl --argstr system x86_64-darwin) $(nix-build https://nixos.org/channels/nixpkgs-unstable/nixexprs.tar.xz -A stdenv.cc.cc --argstr system x86_64-darwin)
/nix/store/xhir85mv5mfwdkbisnx3bl9h6agjmix4-perl-5.28.0
╚═══lib/perl5/5.28.0/darwin-2level/Config_heavy.pl: …rt-5.0.2-dev/include /nix/store/2gdwhdzhy4iwkp7fh8v6gy6nxj1zi9pv-clang-wrapper-5.0.2/resource-ro…
    => /nix/store/2gdwhdzhy4iwkp7fh8v6gy6nxj1zi9pv-clang-wrapper-5.0.2
    ╚═══bin/clang: …k disable=SC2193.[[ "/nix/store/q6wsn9mvlyiwg1risizpx8dwaky70szh-clang-5.0.2/bin/clang" = *++ ]]…
        => /nix/store/q6wsn9mvlyiwg1risizpx8dwaky70szh-clang-5.0.2
```

### PR DESCRIPTION
**Pull request by matthewbauer**
Fixes #46077

### CODE REVIEW

**Review by vcunat — APPROVED** [review]
Seems OK, tested one bootstrapping step on x86_64-linux.  

Disallowing `stdenv.cc.cc` might be better (though it might need some `or` part for bootstrapping), killing two birds with one stone, but it seems good this way.  EDIT: that would need using `disallowedRequisites` instead; I tend to confuse those two :-/

File: pkgs/development/interpreters/perl/default.nix [hunk 1]
```diff
@@ -118,6 +120,7 @@ let
           --replace "${
               if stdenv.cc.cc or null != null then stdenv.cc.cc else "/no-such-path"
             }" /no-such-path \
+          --replace "${stdenv.cc}" /no-such-path \
```
Comment (xeji) [inline pkgs/development/interpreters/perl/default.nix]: This looks  very similar to the substitution above, which seems to do the same job for `x86_64-linux`.
What exactly is the difference between `stdenv.cc` and `stdenv.cc.cc` ?

**Comment by samueldr** [comment]
@matthewbauer I see you added *port to stable*, I don't see any backport.

```
~/tmp/nixpkgs/nixpkgs $ git log master --format='%H %an' --grep="disallow reference to cc" | wc -l
1
~/tmp/nixpkgs/nixpkgs $ git log release-18.09 --format='%H %an' --grep="disallow reference to cc" | wc -l
0
~/tmp/nixpkgs/nixpkgs $ git log staging-18.09 --format='%H %an' --grep="disallow reference to cc" | wc -l
0
```

I'm assuming this would first go through staging-18.09 right? Should this be scheduled not to conflict with the existing stuff in staging-18.09?

**Comment by matthewbauer** [comment]
Yeah forgot to do that. It's not super important but definitely useful.

**Pull request merged**